### PR TITLE
Implementation of Composite Sig

### PR DESCRIPTION
**Pull request by feventura**
This is an implementation of the version 10 [Composite Signatures For Use In Internet PKI](https://datatracker.ietf.org/doc/draft-ounsworth-pq-composite-sigs/).
Current status of this work:
- I'm working on fixing the issues pointed by the Git Actions workflow tests.
- There will be alterations to update this for version 11 when it is released.
- To be precise with the IETF draft, we need to wait for ML-DSA-ipd to be implemented.

### CODE REVIEW

**Comment by baentsch** [comment]
Hi Felipe, very glad to see this contribution to the project. It's now coming at a suboptimal moment though: We're going through an external code security audit and can't accept large commits to "main" until that is complete. Would you consider contributing to a separate branch (say "entrust-composite") for now? That'd also give us opportunity to tune and complete together, e.g. pertaining to testing and integrating the pending addition of ml-dsa?

**Comment by feventura** [comment]
Hello Michael, that is unfortunate but the branch suggestion sounds good.
A bit update, I already made the changes for v11, and I'm working on the git actions issues.

**Comment by baentsch** [comment]
Code audit is through, so we're ready to move forward with this, Felipe. Please mark as Ready for Review as and when you feel this PR is in that stage.

File: oqsprov/oqs_prov.h
Comment (baentsch) [inline oqsprov/oqs_prov.h:181]: don't quite understand this: Where's the key material for composite, then?

File: oqsprov/oqs_prov.h
Comment (baentsch) [inline oqsprov/oqs_prov.h:210]: Is this still accurate (issue seems closed)?

File: oqsprov/oqs_prov.h
Comment (feventura) [reply oqsprov/oqs_prov.h:181]: its in there, but the order is not the same, the hybrid will have Classic then PQ, Composite its the opposite, PQ then classic.
This "unless is composite" is referring to the point that the classic key will not be the first.

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:425]: It's rather obvious given the reference to Entrust' OID space, but in line with #351 would you mind putting a comment at the top of the file referencing yourself (or anyone else taking that responsibility) for keeping these OIDs in line with underlying spec/draft (referencing the latter, too)?

File: scripts/common.py
Comment (baentsch) [inline scripts/common.py:24]: nit: why is the end-of-template comment no longer on its own line?

**Review by baentsch — CHANGES_REQUESTED** [review]
Thanks very much @feventura for all this work! I'm particularly impressed with how well you used the templating mechanism and that all testing is unchanged -- but does its job, looking at CI...

See single comments to get this over the finishing line. I glanced over the code without reviewing every single line, particularly wrt to composite handling (also as I don't have a grip on that spec). Please let me know if you consider this necessary. Otherwise I assume this code has been used successfully in the hackathon testing such as to ascertain interoperability with other implementations: OK assumption?

**Comment by baentsch** [comment]
PS: I'm working right now on a workaround to the Kyber/ML-KEM OID mess as well as the lack of ML-KEM/-DSA support in the distros. Should have a solution tomorrow to rebase on and get CI to "green". That said, the Windows CI failure may warrant your attention, @feventura .

File: oqsprov/oqs_prov.h
Comment (baentsch) [reply oqsprov/oqs_prov.h:181]: Ah, OK. Then the subsequent text no longer is accurate, though: OQS key is at index 0 for composite, then, right?

File: oqsprov/oqs_prov.h
Comment (feventura) [reply oqsprov/oqs_prov.h:210]: I followed the same pattern as the hybrid does, and the setECParams() is still used for the classical key recreate.
I didn't check if the setECParams() workaround is still accurate.
https://github.com/open-quantum-safe/oqs-provider/blob/main/oqsprov/oqsprov_keys.c#L533C1-L535C18

File: oqsprov/oqs_prov.h
Comment (feventura) [reply oqsprov/oqs_prov.h:181]: yes, the i.e. is only valid for hybrids.
I'll update the comment with more accurate text.

File: scripts/common.py
Comment (feventura) [reply scripts/common.py:24]: That was a typo at my end, already fixed it, the change will be on my next commit.

File: oqs-template/generate.yml
Comment (feventura) [reply oqs-template/generate.yml:425]: yes, of course.

**Comment by baentsch** [comment]
One more thought, @feventura : This is such a significant addition to `oqsprovider` that  I'd like to ask you to please consider adding yourself [to the contributors list](https://github.com/open-quantum-safe/oqs-provider?tab=readme-ov-file#team) while addressing the issues (and re-basing to main as CI now should "turn green" also on the PR).

**Comment by baentsch** [comment]
@feventura @SWilson4 Would you mind sharing your thoughts (at least with each other) as to which of your two PRs should be merged first?

My suggestion would be to merge Falcon-padded first such as to allow the Composite's to also support those. Any other opinion good for me, as long as the two of you agree with each other in order to minimize your respective effort re-basing.

**Comment by SWilson4** [comment]
> @feventura @SWilson4 Would you mind sharing your thoughts (at least with each other) as to which of your two PRs should be merged first?
> 
> My suggestion would be to merge Falcon-padded first such as to allow the Composite's to also support those. Any other opinion good for me, as long as the two of you agree with each other in order to minimize your respective effort re-basing.

Makes sense to me. I imagine (?) we also want Falcon-padded to land ASAP so that `oqs-provider` can be released along `liboqs`.

**Comment by feventura** [comment]
> @feventura @SWilson4 Would you mind sharing your thoughts (at least with each other) as to which of your two PRs should be merged first?
> 
> My suggestion would be to merge Falcon-padded first such as to allow the Composite's to also support those. Any other opinion good for me, as long as the two of you agree with each other in order to minimize your respective effort re-basing.

Go ahead, the composite draft dropped Falcon in https://github.com/EntrustCorporation/draft-ounsworth-composite-sigs/pull/138 so it should not impact the composite implementation.

**Comment by baentsch** [comment]
@feventura : We're now closing in on the release: Is this ready from your perspective? Should I give it a final review?

File: oqsprov/oqs_prov.h
Comment (baentsch) [reply oqsprov/oqs_prov.h:212]: ACK, so I didn't. So out of scope for this PR.

**Comment by feventura** [comment]
@baentsch Yes, I fixed the problems with the windows CI

**Review by baentsch — APPROVED** [review]

File: README.md
Comment (feventura) [reply README.md:71]: I can add it, there is one other mention of the draft spec in generate.yml:429.
There is no mention of what spec of the draft in any other place.

Comment (baentsch) [reply README.md:71]: That would be good, I think (for those few people reading a README :).

That reminds me, @dstebila, is there a comparable spec documenting hybrids that we may add in parallel? My rapidly decaying brain cells tell me there was something somewhere, but I fail to locate it....

Comment (dstebila) [reply README.md:71]: For hybrid KEMs in TLS, yes: https://datatracker.ietf.org/doc/draft-ietf-tls-hybrid-design/

For hybrids signatures in TLS and X.509, I don't think so, we're currently just doing something of our design.

Comment (feventura) [reply README.md]: I added the link for the composite draft in the README. (:

Comment (baentsch) [reply README.md]: Will add the hybrid link in a follow-up PR before release. Thanks for that, @dstebila

**Comment by feventura** [comment]
> @feventura Thanks again for all this work. One nit regarding documentation that you may want to fix while re-basing. Would it be possible to also put this all into one commit such as to not clutter the commit history?
> 
> Last item addressing @thb-sb : Would you have time to give this PR a second pair of "review" eyes prior to merge?

Thanks! (: 
I'm doing the rebase today and solving the conflicts.
After @thb-sb give the feedback from the "review" I'll squash my commits into a single one.
I'll wait for this review in case I'll need to make further changes.

File: oqsprov/oqs_encode_key2any.c
Comment (feventura) [reply oqsprov/oqs_encode_key2any.c:533]: This is done for non-composite.
There is a if statement for that in line 514.
But I agree with you, seems like a leak.

File: oqsprov/oqs_encode_key2any.c
Comment (feventura) [reply oqsprov/oqs_encode_key2any.c:815]: No, there isn't any reason, I'm adding it just in case the malloc fails.

File: oqsprov/oqs_encode_key2any.c
Comment (feventura) [reply oqsprov/oqs_encode_key2any.c:850]: I'm supposing you are speaking about the OPENSSL_cleanse.
I cannot free buf at that point because buf is used inside p8inf_internal, so if I free it when I free p8inf_internal I would get a double free error.
So I'm cleansing it to remove the pkey data from memory and on the next line I'm doing [PKCS8_PRIV_KEY_INFO_free](https://www.openssl.org/docs/manmaster/man3/PKCS8_PRIV_KEY_INFO_free.html) to free the p8inf_internal which include the buf.

File: oqsprov/oqs_encode_key2any.c
Comment (feventura) [reply oqsprov/oqs_encode_key2any.c:823]: I accept it because at first glance it seems fine, but then I went to test it and remembered why I did that.
Some cases of the RSA keys the size of the key will not match the size on oqsxkey->privkeylen_cmp, this value represent the maximum size for the RSA, so I cant copy oqsxkey->privkeylen_cmp because it will cause overflow in some cases.

File: oqsprov/oqsprov_keys.c
Comment (feventura) [reply oqsprov/oqsprov_keys.c:1]: For public key is simple as that, private is more challenging, because RSA privkey will not always match the maximum size, and EC can exist without ECParameters and Pubkey.

File: oqs-template/oqsprov/oqsprov.c/encoding_patching.fragment
Comment (baentsch) [reply oqs-template/oqsprov/oqsprov.c/encoding_patching.fragment:14]: Good suggestion, but as you said, out of scope here: Would you mind creating an issue about this, @thb-sb ?

File: oqsprov/oqsprov_keys.c
Comment (baentsch) [reply oqsprov/oqsprov_keys.c:1]: So what about adding these thoughts at least as comments such as to be able to understand the logic and arithmetic in the code?

Comment (feventura) [reply oqsprov/oqsprov_keys.c:1]: That makes a lot of sense, Ill go in the places I know I had to do some logic that can be hard to understand at first glance.
I'll add comments to help guide who is reading the code.

File: oqsprov/oqs_sig.c
Comment (feventura) [reply oqsprov/oqs_sig.c:408]: Makes sense, I removed the free from this and all other cases that I'm doing this checkup

File: oqsprov/oqs_sig.c
Comment (feventura) [reply oqsprov/oqs_sig.c:449]: No we can't, name is used a lot inside this loop.

File: oqsprov/oqs_sig.c
Comment (feventura) [reply oqsprov/oqs_sig.c:751]: In this final_tbs, after its been allocated in there, is freed on every error before going to endverify.

File: oqsprov/oqs_sig.c
Comment (feventura) [reply oqsprov/oqs_sig.c:438]: Thanks, I missed adding this frees on sign as I did on verify.

File: oqsprov/oqsprov.c
Comment (feventura) [reply oqsprov/oqsprov.c:1178]: added.

File: oqsprov/oqsprov_keys.c
Comment (feventura) [reply oqsprov/oqsprov_keys.c:705]: added.

File: oqsprov/oqsprov.c
Comment (feventura) [reply oqsprov/oqsprov.c:1186]: I made the change and added the error handling.

**Comment by baentsch** [comment]
Err, IMO this is tested--at least the new algs show up in CI... Testament to the quality of the test framework :-)

**Comment by baentsch** [comment]
Just re-started CI to re-confirm...

File: oqsprov/oqsprov_keys.c
Comment (baentsch) [reply oqsprov/oqsprov_keys.c:1]: >  The composite signature is not a trivial feature, and it adds a lot of code. Since this is major, I think it would be nice to have it under its own file (or set of files).
oqsprov/oqsprov.c and oqsprov/oqsprov_keys.c are already large.
> 
> I also suggest that we split your code into smaller functions, in order to improve reading and to factor some bit of code.

I tend to agree, @thb-sb -- but you also state that the code/files even before this PR is large (mea culpa)... So what about the suggestion to create an issue for doing a "general" refactoring (also of the existing "hybrid" code) but let this PR land first to not "overload" it with tasks (that also would sensibly done to other parts of the code)? Secondary benefit of this approach: If @feventura does not volunteer to do it, the person doing the refactoring gets a much better grip on the code than is possible by reviewing alone.

**Comment by baentsch** [comment]
@thb-sb , fyi, CI re-run confirms both command line and API-level test of (also) composite sigs to have occurred and passed.

**Comment by baentsch** [comment]
So, glancing over all comments I think we've only got things left that are not "Composite-specific", agree, @thb-sb ? So would it be good to ask @feventura for a final squash to then commit this to "main"? Or did I overlook something? I'd like to merge this to then move towards a release soon...

**Comment by feventura** [comment]
@baentsch @thb-sb I implemented the changed requested by Thomas and I solved the DCO check by rebasing to add signature to the old commits.
You should be able to squash all the commits before accepting the merge.
Is there some other changes you would like me to do?

**Comment by baentsch** [comment]
Thanks for the "good-to-go" message, @feventura . I think @thb-sb has some concerns still, but may look which ones are essential before landing and which ones should be covered as issues for future resolution. Regarding the merge, may I ask you to squash all into one commit such as for you to decide which commit message(s) to retain for "posteriority" -- I suppose not all 164 commit messages are relevant, right? If I were squashing this, I'd delete _all_ messages -- but this is possibly not what you want (?).

**Comment by baentsch** [comment]
Thanks for the review, @thb-sb ! @feventura it looks like `asan` is unhappy about the code now (2 tests crashing): Could you please take a look?

**Comment by feventura** [comment]
> > Thanks for the review, @thb-sb ! @feventura it looks like `asan` is unhappy about the code now (2 tests crashing): Could you please take a look?
> 
> Regarding these ASan tests, I've noticed something quite unusual while working on #374, which makes me think that it's unrelated to @feventura changes: I also had totally random failure on this job. Since everything is compiled and linked against ASan, and since we have `--output-on-failure`, in case of a memory leak or a memory corruption, we should see an ASan report.
> 
> But in that case (and same for my PR), we don't have any ASan report being displayed before the segfault. This is why I've added `--repeat until-pass:5` here: https://github.com/thb-sb/oqs-provider/blob/3b69bbe728819ad9b325580b79d9c9a672ae366e/.github/workflows/linux.yml#L149-L152
> 
> I think GitHub has changed their runner configuration recently, and something has been broken here. I've run these tests a million times locally using [`act`](https://github.com/nektos/act) (to ensure that they were running under the very same environment as GitHub action's). I couldn't managed to reproduce a single crash.
> 
> My guess is that since ASan is virtually allocating the whole address space to catch any memory corruption ops, someone may see this as a process requiring too much memory, although this is purely virtual.

Yes I found weird that ASan reported as SegFault but I have not extra information about it. I also tested locally several times locally using [`act`](https://github.com/nektos/act) and it didn`t crashed once.

What I noticed is that the GitHub actions in this PR failed due to oqs_signatures and oqs_endecode. On my fork it passed successfully when I pushed Saturday, so I repeated on my fork today and it failed due to oqs_tlssig and oqs_endecode, which also point to the totally random failure on this job.

Regardless I will run [`act`](https://github.com/nektos/act) locally several times to see today if I can reproduce it.

**Comment by baentsch** [comment]
> added --repeat until-pass:5

worthwhile trying also in this PR, @feventura? Clearly just a workaround, but better than wasting more effort on something we may not be able to help

File: oqsprov/oqs_encode_key2any.c
Comment (baentsch) [reply oqsprov/oqs_encode_key2any.c:533]: Created #375 to track

File: oqsprov/oqs_sig.c
Comment (baentsch) [reply oqsprov/oqs_sig.c:1]: Created #375 to track

File: oqsprov/oqsprov_keys.c
Comment (baentsch) [reply oqsprov/oqsprov_keys.c:1]: Created #375 to track

**Comment by baentsch** [comment]
@feventura @thb-sb Thanks again for your diligence & support to work through this PR. If I don't hear objections, I will merge later today, squashing all commits into one just saying "Adding CompositeSig support".

**Comment by baentsch** [comment]
Merged. Thanks very much for the contribution, @feventura !

**Comment by bhess** [comment]
> Yes, I found weird that ASan reported as SegFault but I have not extra information about it.

I had the same issue with another project starting last week. Seems to be an issue with the [github runner image](https://github.com/actions/runner-images/issues/9491). There's a fix that appears to be deployed soon.

**Pull request merged**
